ccl/changefeedccl: TestChangefeedTestTimesOut failed

ccl/changefeedccl.TestChangefeedTestTimesOut [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/10308488?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/10308488?buildTab=artifacts#/) on release-23.1 @ [f462a0e04526d90b004b6865938587a1bec28c21](https://github.com/cockroachdb/cockroach/commits/f462a0e04526d90b004b6865938587a1bec28c21):


```
=== RUN   TestChangefeedTestTimesOut
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/c38f6012ef4a0b1e63cc5ae80f5f2e37/logTestChangefeedTestTimesOut3737325401
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestChangefeedTestTimesOut
    dir.go:41: unlinkat /artifacts/tmp/_tmp/c38f6012ef4a0b1e63cc5ae80f5f2e37/TestChangefeedTestTimesOut1295551541/6: directory not empty
    changefeed_test.go:8070: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/c38f6012ef4a0b1e63cc5ae80f5f2e37/logTestChangefeedTestTimesOut3737325401
--- FAIL: TestChangefeedTestTimesOut (7.49s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #95139 ccl/changefeedccl: TestChangefeedTestTimesOut failed [C-test-failure O-robot T-cdc branch-master]
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedTestTimesOut.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/changefeedccl/changefeed_test.go
+++ b/pkg/ccl/changefeedccl/changefeed_test.go
@@ -8168,7 +8168,11 @@ func TestChangefeedTestTimesOut(t *testing.T) {
 		sqlDB := sqlutils.MakeSQLRunner(s.DB)
 		sqlDB.Exec(t, `CREATE TABLE foo (a INT PRIMARY KEY)`)
 		nada := feed(t, f, "CREATE CHANGEFEED FOR foo WITH resolved='100ms'")
-		defer closeFeed(t, nada)
+		defer func() {
+			// close could return an error due to the race in withTimeout function
+			// which cancels the job.
+			_ = nada.Close()
+		}()
 
 		expectResolvedTimestamp(t, nada) // Make sure feed is running.
 

--- a/pkg/ccl/changefeedccl/helpers_test.go
+++ b/pkg/ccl/changefeedccl/helpers_test.go
@@ -718,7 +718,6 @@ func expectErrCreatingFeed(
 }
 
 func closeFeed(t testing.TB, f cdctest.TestFeed) {
-	t.Helper()
 	if err := f.Close(); err != nil {
 		t.Fatal(err)
 	}

--- a/pkg/testutils/dir.go
+++ b/pkg/testutils/dir.go
@@ -38,7 +38,10 @@ func TempDir(t testing.TB) (string, func()) {
 	}
 	cleanup := func() {
 		if err := os.RemoveAll(dir); err != nil {
-			t.Error(err)
+			// Temp dir cleanup may race with process shutdown, where
+			// some process is writing to dir, just as we try to shut down.
+			// This should not fail the test.
+			t.Logf("encountered error %s while removing temp dir %s", err, dir)
 		}
 	}
 	return dir, cleanup

